File: app/src/main/java/com/github/anrimian/musicplayer/ui/playlist_screens/create/CreatePlayListDialogFragment.java
package com.github.anrimian.musicplayer.ui.playlist_screens.create;

import static com.github.anrimian.musicplayer.ui.utils.AndroidUtils.setSoftInputVisible;

import android.app.AlertDialog;
import android.app.Dialog;
import android.os.Bundle;
import android.text.InputType;
import android.view.LayoutInflater;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.github.anrimian.musicplayer.R;
import com.github.anrimian.musicplayer.databinding.DialogCommonInputBinding;
import com.github.anrimian.musicplayer.di.Components;
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList;
import com.github.anrimian.musicplayer.ui.common.error.ErrorCommand;
import com.github.anrimian.musicplayer.ui.utils.OnCompleteListener;

import moxy.MvpAppCompatDialogFragment;
import moxy.presenter.InjectPresenter;
import moxy.presenter.ProvidePresenter;

public class CreatePlayListDialogFragment extends MvpAppCompatDialogFragment
        implements CreatePlayListView {

    @InjectPresenter
    CreatePlayListPresenter presenter;

    private EditText etPlayListName;
    private TextView tvError;
    private TextView tvProgress;
    private ProgressBar progressBar;

    private Button btnCreate;

    @Nullable
    private OnCompleteListener<PlayList> onCompleteListener;

    @ProvidePresenter
    CreatePlayListPresenter providePresenter() {
        return Components.getAppComponent().createPlayListsPresenter();
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        DialogCommonInputBinding viewBinding = DialogCommonInputBinding.inflate(
                LayoutInflater.from(requireActivity())
        );
        etPlayListName = viewBinding.etInput;
        tvError = viewBinding.tvError;
        tvProgress = viewBinding.tvProgress;
        progressBar = viewBinding.progressBar;

        View view = viewBinding.getRoot();

        AlertDialog dialog = new AlertDialog.Builder(getActivity())
                .setTitle(R.string.create_playlist)
                .setPositiveButton(R.string.create, null)
                .setNegativeButton(R.string.cancel, (dialog1, which) -> {})
                .setView(view)
                .create();
        setSoftInputVisible(dialog.getWindow());
        dialog.show();

        etPlayListName.setImeOptions(EditorInfo.IME_ACTION_DONE);
        etPlayListName.setRawInputType(InputType.TYPE_CLASS_TEXT);
        etPlayListName.setOnEditorActionListener((v, actionId, event) -> {
            onCompleteButtonClicked();
            return true;
        });
        etPlayListName.requestFocus();

        btnCreate = dialog.getButton(AlertDialog.BUTTON_POSITIVE);
        btnCreate.setOnClickListener(v -> onCompleteButtonClicked());

        return dialog;
    }

    @Override
    public void showProgress() {
        btnCreate.setEnabled(false);
        etPlayListName.setEnabled(false);
        tvError.setVisibility(View.GONE);
        tvProgress.setVisibility(View.VISIBLE);
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void showInputState() {
        btnCreate.setEnabled(true);
        etPlayListName.setEnabled(true);
        tvError.setVisibility(View.GONE);
        tvProgress.setVisibility(View.GONE);
        progressBar.setVisibility(View.GONE);
    }

    @Override
    public void showError(ErrorCommand errorCommand) {
        btnCreate.setEnabled(true);
        etPlayListName.setEnabled(true);
        tvProgress.setVisibility(View.GONE);
        progressBar.setVisibility(View.GONE);
        tvError.setVisibility(View.VISIBLE);
        tvError.setText(getString(R.string.create_playlist_error_template, errorCommand.getMessage()));
    }

    @Override
    public void onPlayListCreated(PlayList playList) {
        if (onCompleteListener != null) {
            onCompleteListener.onComplete(playList);
        }
        dismissAllowingStateLoss();
    }

    public void setOnCompleteListener(@Nullable OnCompleteListener<PlayList> onCompleteListener) {
        this.onCompleteListener = onCompleteListener;
    }

    private void onCompleteButtonClicked() {
        presenter.onCompleteInputButtonClicked(etPlayListName.getText().toString());
    }
}

File: app/src/main/java/com/github/anrimian/musicplayer/ui/playlist_screens/create/CreatePlayListDialogFragment.kt
package com.github.anrimian.musicplayer.ui.playlist_screens.create

import android.app.AlertDialog
import android.app.Dialog
import android.os.Bundle
import android.text.InputType
import android.view.LayoutInflater
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.Button
import com.github.anrimian.musicplayer.R
import com.github.anrimian.musicplayer.databinding.DialogCommonInputBinding
import com.github.anrimian.musicplayer.di.Components
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList
import com.github.anrimian.musicplayer.ui.common.error.ErrorCommand
import com.github.anrimian.musicplayer.ui.utils.AndroidUtils
import moxy.MvpAppCompatDialogFragment
import moxy.ktx.moxyPresenter

class CreatePlayListDialogFragment : MvpAppCompatDialogFragment(), CreatePlayListView {

    private val presenter by moxyPresenter { Components.getAppComponent().createPlayListsPresenter() }
    
    private lateinit var binding: DialogCommonInputBinding
    
    private lateinit var btnCreate: Button
    
    private var onCompleteListener: ((PlayList) -> Unit)? = null

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        binding = DialogCommonInputBinding.inflate(LayoutInflater.from(requireActivity()))

        val dialog = AlertDialog.Builder(activity)
            .setTitle(R.string.create_playlist)
            .setPositiveButton(R.string.create, null)
            .setNegativeButton(R.string.cancel) { _, _ -> }
            .setView(binding.root)
            .create()
        
        AndroidUtils.setSoftInputVisible(dialog.window)
        dialog.show()
        
        binding.etInput.imeOptions = EditorInfo.IME_ACTION_DONE
        binding.etInput.setRawInputType(InputType.TYPE_CLASS_TEXT)
        binding.etInput.setOnEditorActionListener { _, _, _ ->
            onCompleteButtonClicked()
            true
        }
        binding.etInput.requestFocus()
        
        btnCreate = dialog.getButton(AlertDialog.BUTTON_POSITIVE)
        btnCreate.setOnClickListener { onCompleteButtonClicked() }
        
        return dialog
    }

    override fun showProgress() {
        btnCreate.isEnabled = false
        binding.etInput.isEnabled = false
        binding.tvError.visibility = View.GONE
        binding.tvProgress.visibility = View.VISIBLE
        binding.progressBar.visibility = View.VISIBLE
    }

    override fun showInputState() {
        btnCreate.isEnabled = true
        binding.etInput.isEnabled = true
        binding.tvError.visibility = View.GONE
        binding.tvProgress.visibility = View.GONE
        binding.progressBar.visibility = View.GONE
    }

    override fun showError(errorCommand: ErrorCommand) {
        btnCreate.isEnabled = true
        binding.etInput.isEnabled = true
        binding.tvProgress.visibility = View.GONE
        binding.progressBar.visibility = View.GONE
        binding.tvError.visibility = View.VISIBLE
        binding.tvError.text = getString(R.string.create_playlist_error_template, errorCommand.message)
    }

    override fun onPlayListCreated(playList: PlayList) {
        onCompleteListener?.invoke(playList)
        dismissAllowingStateLoss()
    }

    fun setOnCompleteListener(onCompleteListener: (PlayList) -> Unit) {
        this.onCompleteListener = onCompleteListener
    }

    private fun onCompleteButtonClicked() {
        presenter.onCompleteInputButtonClicked(binding.etInput.text.toString())
    }
}